delta-lake/src/main/java/org/apache/iceberg/delta/BaseSnapshotDeltaLakeTableAction.java

```diff
@@ -0,0 +1,403 @@
+    Iterator<VersionLog> versionLogIterator =
+        deltaLog.getChanges(
+            deltaStartVersion, false // not throw exception when data loss detected
+            );
+    while (versionLogIterator.hasNext()) {
+      VersionLog versionLog = versionLogIterator.next();
+      commitDeltaVersionLogToIcebergTransaction(versionLog, icebergTransaction);
+    }
+    Snapshot icebergSnapshot = icebergTransaction.table().currentSnapshot();
+    long totalDataFiles =
+        icebergSnapshot != null
+            ? Long.parseLong(icebergSnapshot.summary().get(SnapshotSummary.TOTAL_DATA_FILES_PROP))
+            : 0;
+
+    icebergTransaction.commitTransaction();
+    LOG.info(
+        "Successfully created Iceberg table {} from delta lake table at {}, total data file count: {}",
+        newTableIdentifier,
+        deltaTableLocation,
+        totalDataFiles);
+    return new BaseSnapshotDeltaLakeTableActionResult(totalDataFiles);
+  }
+
+  private Schema convertDeltaLakeSchema(io.delta.standalone.types.StructType deltaSchema) {
+    Type converted =
+        DeltaLakeDataTypeVisitor.visit(deltaSchema, new DeltaLakeTypeToType(deltaSchema));
+    return new Schema(converted.asNestedType().asStructType().fields());
+  }
+
+  private PartitionSpec getPartitionSpecFromDeltaSnapshot(Schema schema) {
+    List<String> partitionNames = deltaLog.snapshot().getMetadata().getPartitionColumns();
+    if (partitionNames.isEmpty()) {
+      return PartitionSpec.unpartitioned();
+    }
+
+    PartitionSpec.Builder builder = PartitionSpec.builderFor(schema);
+    for (String partitionName : partitionNames) {
+      builder.identity(partitionName);
+    }
+    return builder.build();
+  }
+
+  /**
+   * Iterate through the {@code VersionLog} to determine the update type and commit the update to
+   * the given {@code Transaction}.
+   *
+   * <p>There are 3 cases:
+   *
+   * <p>1. AppendFiles - when there are only AddFile instances (an INSERT on the table)
+   *
+   * <p>2. DeleteFiles - when there are only RemoveFile instances (a DELETE where all the records of
+   * file(s) were removed)
+   *
+   * <p>3. OverwriteFiles - when there are a mix of AddFile and RemoveFile (a DELETE/UPDATE)
+   *
+   * @param versionLog the delta log version to commit to iceberg table transaction
+   * @param transaction the iceberg table transaction to commit to
+   */
+  private void commitDeltaVersionLogToIcebergTransaction(
+      VersionLog versionLog, Transaction transaction) {
+    List<Action> dataFileActions;
+    if (versionLog.getVersion() == deltaStartVersion) {
+      // The first version log is a special case, since it contains the initial table state.
+      // we need to get all dataFiles from the corresponding delta snapshot to construct the table.
+      dataFileActions =
+          deltaLog.getSnapshotForVersionAsOf(deltaStartVersion).getAllFiles().stream()
+              .map(addFile -> (Action) addFile)
+              .collect(Collectors.toList());
```

Comment:
It is logically reasonable to extract the datafiles from the snapshot for the start version only instead of from the version log, since there may be versions already deleted from the table. 

Also, doing this also allow us to add options to let users choose which version to start snapshot.